### PR TITLE
Add syntax for non maximal implicit arguments

### PR DESCRIPTION
**Pull request by SimonBoulier**
**Kind:** feature.

Add syntax for non maximal implicit arguments as suggested in #7767.

The syntax used is: `[x : A]`, `[x]`, `` `[A]`` as suggested by the command `Arguments`. It does not seems to conflict too much (no conflict in the stdlib).

E.g.
```
    Definition foo [n] m := n + m.
-> n is implicit, not maximally inserted (printed [n] by About)
```

What is the process to decide whether it is a good idea?
I personally feel non maximal implicit arguments useful.
@herbelin, if there is a consensus on it I can do more testing and add documentation.

We could also discuss the behavior of non maximal trailing arguments: as of today there are transformed as maximal ones with a "deprecated" warning.
E.g.
```
    Definition foo n [m] := n + m.
-> m is implicit, maximally inserted (printed {m} by About)
```
This is because a trailing non max implicit won't never be inserted.
They could also be transformed as non implicit args, maybe it is closer to the expected behavior...

- [x] Added / updated test-suite

- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

**Comment by herbelin** [comment]
> What is the process to decide whether it is a good idea?
I personally feel non maximal implicit arguments useful.
@herbelin, if there is a consensus on it I can do more testing and add documentation.

As far as I remember, little people gave an opinion. @maximedenes, who considered implementing the feature at some time, told he stopped precisely because he was not able to evaluate whether this was useful or not, but, as far as I understand, he was not opposed, on the contrary, since it started to develop it. My position is that either we suppress non-maximal implicit arguments completely - but this seems unrealistic since many developments rely on them -, or we accept them also in manual mode. The possible worry is that it would prevent the syntax `[...]` to be used for something else, but to what this syntax could be useful if not for non-maximal implicit arguments. So, my answer is that it is meaningful.

Is there a need for more discussion? @mattam82, what do you think? If noone has an objection to raise for one or another reason, I guess we can consider that there is a consensus.

Then a next step would be to add the syntax `{x}%delim` and `[x]%delim` in binders. The pros are that it would improve the consistent with `Arguments`. The cons is that it might look  a bit cryptic to decipher for casual readers. It may also be not so useful in practice, as most of the time, the type of the variable enforces a scope.

**Comment by SkySkimmer** [comment]
>Then a next step would be to add the syntax {x}%delim and [x]%delim in binders.

That would read somewhat close to opening the scope for the type of the binder. Not quite ambiguous but still somehow questionable.

**Comment by herbelin** [comment]
That's right... That should be on ident name then. Not as nice, but also supported by `Arguments`.

**Comment by Zimmi48** [comment]
I'm in favor. Also for the choice of the warning for trailing arguments. Is that really a "deprecated" warning though? It shouldn't be one. Deprecated means that it's going to stop being supported.

**Comment by SimonBoulier** [comment]
> Is that really a "deprecated" warning though? It shouldn't be one. Deprecated means that it's going to stop being supported.

Indeed it is weird, it was as this in the code but I can change it.

**Comment by mattam82** [comment]
Personally, I would probably use such a feature, especially for lemmas where one might want to use the fully explicit form sometimes, and other times only the "non-maximal implicit" version, instantiating non-dependent hypotheses directly. It also makes the result of the automatic inference of implicits more direct to see. I'm wondering if people developing libraries recently still use non-maximal implicits much. @robbertkrebbers or @charguer for example ? I'm not so sure about mathcomp: does it recommends the use of prenex non-maximal implicits only? My point is that if non-maximal implicits are here to stay then we can see this as a welcome improvement of the situation. BTW, a bit of history: when I added the explicit maximal syntax `{ }` there was contention with using `[ ]` because that was being used in the ICC* (implicit calculus of constructions) branch of Coq, which in the end was not merged. I'm sure a different binding annotation could be used for that though, e.g `(x :* I)` or somesuch.

**Comment by charguer** [comment]
Since you ask my point of view on this topic...

- In the current system, it is quite often the case that you want arguments to be implicit but not maximally-inserted. Basically, most lemmas about polymorphic functions.
```
Lemma fmap_extens : forall A B `{Inhab B} (m1 m2:fmap A B), 
 dom m1 = dom m2 ->
 (forall x, dom m1 x -> read m1 x = read m2 x) -> 
 m1 = m2
```
you typically want to have A and B implicit, but certainly not maximally inserted, else by default "fmap_extens" gets an arbitrary instantiation of the types, typically `(B:=unit)` because `unit` is the global first instance for an inhabited type.
Whether you want `m1` and  `m2` to be implicit or not is a matter of taste; both have pros and cons.

Now, what syntax for specifying which arguments are implicit or maximally-implicit?

- For definitions, additional syntax such as `[x:A]` or `{x:A}` tend to make specifications harder to read for non-experts; this is not a good thing, specifications should be readable even by people who don't master Coq.

- A possibly useful feature would to make the "Arguments" command have the definition name be optional (or possibly a "_" argument), to indicate that the command applies to the immediately preceeding top-level definition. Yet maintaining a list of named arguments with the "Arguments" command is somewhat heavy and a pain while maintaining scripts.

Perhaps, better is to avoid the need for implicit arguments altogether.

- For lemmas, if using the "first-match instantiation mode for application" (as provided by TLC tactics), it does not matter at all which arguments implicit or not or maximally-inserted. (See TLC documentation or ask me for more details.)

- If this "first-match" feature was available for all terms (e.g. "#foo x y" is equivalent to "@foo _ .. _ x _ .. _ y" for the appopriate number of underscores, then I believe you would no longer need to care for arguments that are implicit but not maximally-implicit. In other words, staying with the current features of having "normal arguments" and "maximally-implicit arguments" would be sufficient. 

- Remark : the maximally-implicit arguments are quite useful for, e.g. ̀nil : forall {A:Type}, list A` and other polymorphic constructors without arguments. I suspect these case are rare enough that you could deal with them using the `Arguments` command instead of the ̀{A}` syntax. Thus, perhaps one could live even without the syntax `{A:Type}` in quantifiers, satisfying my requirement for "accessible specifications".

**Comment by Zimmi48** [comment]
> For definitions, additional syntax such as `[x:A]` or `{x:A}` tend to make specifications harder to read for non-experts; this is not a good thing, specifications should be readable even by people who don't master Coq.

Meh. I'm pretty sure brackets surrounding arguments are pretty easy to ignore if you don't master Coq. OTOH, having arguments made automatically implicit is much harder to grasp because you need to have the algorithm in your head.

**Comment by herbelin** [comment]
@charguer: I share the view that specifications should ideally be accessible to non-experts as much as possible. (On the other side, `{ ... }` is light and tends to fall in a common practice, shared by Agda, Lean, ...)

Can you tell more about "first-match instantiation mode for application" (is it based on the number of arguments, or based on type? how is disambiguation addressed)? In particular, the mathematical practice quite flexibly manages optional arguments and, morally, we should be able to mechanize this flexibility.

**Comment by SimonBoulier** [comment]
The PR is ready for reviews. The changes make Elpi and Quichick failing, I can do the patches after merge.

**Comment by SkySkimmer** [comment]
>The PR is ready for reviews. The changes make Elpi and Quichick failing, I can do the patches after merge.

You're supposed to do it before using the overlay system https://github.com/coq/coq/blob/master/dev/ci/user-overlays/README.md

**Comment by SimonBoulier** [comment]
Remark: the changes implied in the test-suite were marked deprecated since Coq 8.10.
For example `Arguments nil [_].` raised:
```
Warning: Argument number 0 is a trailing implicit so must be maximal
[set-maximal-deprecated,deprecated]
```
(and now it raises an error).

**Review by JasonGross — CHANGES_REQUESTED** [review]
Please add mention of the new warning/error to the changelog

**Review by jfehrle — CHANGES_REQUESTED** [review]
Mostly minor wording suggestions, but please be sure to correct the grammar description in the documentation.

File: interp/impargs.ml
Comment (SimonBoulier) [reply interp/impargs.ml]: `{ contains { so I think it's ok as is.

**Comment by SimonBoulier** [comment]
Thanks @JasonGross  and @jfehrle  for your comments. I'll adress them after holidays.
@SkySkimmer thanks, I can do that.

**Review by jfehrle — APPROVED** [review]
Reviewed for wording, looks good.

**Comment by gares** [comment]
There is a detail about the new error on trailing non-maximal implicits that is worth discussing.
In particular it does not seem true that trailing non-maximal implicit are never inserted.

For example the line failing in the OOTHM is about `th_bboxP` that has two arguments, but its type `reflect` coerces to funclass, hence it can be applied to more arguments than what your new error thinks about.
https://github.com/math-comp/odd-order/blob/4975ca9d873c48526efa56de34279d0f302389c4/theories/PFsection3.v#L255-L263
See for example how it is used here
  https://github.com/math-comp/odd-order/blob/4975ca9d873c48526efa56de34279d0f302389c4/theories/PFsection3.v#L767
that under the hood means `coercion (th_bboxP _ _) argument`. Now, making the arguments maximally implicit works in this case but changes what `th_bboxP` means (since you get the `_` even if there is no argument). I don't know if it is a problem in this specific case (CC @ggonthier probably the author of that choice of implicit arguments).

**Review by maximedenes — DISMISSED** [review]
I like this PR, thanks! However, it currently does two things at once, as witnessed by the two entries in the changelog. Please separate it in two.

**Comment by SimonBoulier** [comment]
@gares Thanks for you remark, I had the same thinking.
But in fact, I think that I don't change anything. The implicit arguments were already declared maximally inserted automatically (because they are trailing), and `th_bboxP` already meant `th_bboxP _ _`. So I don't change the behavior, I only make it explicit. (And the coercion doesn't bring anything new.)
Am I correct?

**Comment by maximedenes** [comment]
> But in fact, I think that I don't change anything. The implicit arguments were already declared maximally inserted automatically (because they are trailing), and `th_bboxP` already meant `th_bboxP _ _`. So I don't change the behavior, I only make it explicit. (And the coercion doesn't bring anything new.)
> Am I correct?

That seems right, indeed. But I'm wondering: were the examples that you fixed triggering the warning? It seems surprising that nobody noticed it.

**Comment by SimonBoulier** [comment]
@maximedenes (about splitting the PR) I understand it would have been a better choice. But is it really necessary while the PR is almost ready? Not so little work...

**Comment by SimonBoulier** [comment]
> > 
> 
> That seems right, indeed. But I'm wondering: were the examples that you fixed triggering the warning? It seems surprising that nobody noticed it.

Are contribs supposed not trigger any warning? It seems to me that I saw many of them while compiling several contribs.

**Comment by maximedenes** [comment]
> I understand it would have been a better choice. But is it really necessary while the PR is almost ready? Not so little work...

Sorry, but we had many problems with PRs mixing several changes, I think we should really try to be less sloppy about it.

In fact, if you look at the title of this PR and the overlays, they are completely unrelated...

If you think it is too much work, I can try to do it.

**Comment by SimonBoulier** [comment]
I splitted the PR in three: #11368, #11369 and this one. For me it's ready modulo CI.

**Comment by SimonBoulier** [comment]
@maximedenes and @JasonGross Are your requests satisfied? Can you approve the PR?

File: doc/changelog/02-specification-language/11235-non_maximal_implicit.rst
Comment (JasonGross) [inline doc/changelog/02-specification-language/11235-non_maximal_implicit.rst]: Does this mean that the syntax is not valid underneath forall/fun?  If so, why this restriction?  If not, the changelog should not imply that it's only valid for top-level definitions.

File: doc/sphinx/language/gallina-extensions.rst
Comment (JasonGross) [inline doc/sphinx/language/gallina-extensions.rst]: Documentation of this warning should not be removed, but should instead be changed to be documenting an error.

**Review by JasonGross — DISMISSED** [review]
Sorry for the delay in rereviewing, and thanks for the ping

File: doc/changelog/02-specification-language/11235-non_maximal_implicit.rst
Comment (SimonBoulier) [reply doc/changelog/02-specification-language/11235-non_maximal_implicit.rst]: You are right, it works everywhere!

File: doc/sphinx/language/gallina-extensions.rst
Comment (SimonBoulier) [reply doc/sphinx/language/gallina-extensions.rst]: This is the case but I moved it, see  lines 1586 to 1592:
```
.. exn:: Argument @ident is a trailing implicit, so it can't be declared non maximal. Please use %{ %} instead of [ ].

   For instance:

   .. coqtop:: all fail

      Fail Definition double [n] := n + n.
```

**Comment by gares** [comment]
This PR needs an assignee, the ideal candidate would be someone from @coq/parsing-maintainers

**Comment by gares** [comment]
BTW, I did a mistake and merged the overlay in coq-elpi (CC @coq/ci-maintainers). Do you want me to revert it right now?

File: doc/changelog/02-specification-language/11235-non_maximal_implicit.rst
Comment (maximedenes) [reply doc/changelog/02-specification-language/11235-non_maximal_implicit.rst]: I'm not sure it works properly everywhere. Let's confirm it is the case before extending the new syntax to local binders.

**Comment by ejgallego** [comment]
> BTW, I did a mistake and merged the overlay in coq-elpi (CC @coq/ci-maintainers). Do you want me to revert it right now?

Please @gares , indeed the CI for elpi is broken due to this.

**Comment by gares** [comment]
Done, here the new PR/overlay: https://github.com/LPCIC/coq-elpi/pull/113

**Comment by SimonBoulier** [comment]
@JasonGross and @maximedenes I added some tests in the test-suite and indeed I caught a bug. Now, local non max implicit arguments seem to work well.

There is only a small thing which is not perfect: we do not forbid implicit arguments which are both local, trailing and non max. E.g.
```
Definition F (id: forall [A] [x : A], A) := id.
```
It may fail with `Argument A is a trailing implicit, so it can't be declared non maximal. Please use
{ } instead of [ ].`. Do you think it is important? It does not introduce any bug and seems complicated to be handled nicely in the code.

**Comment by herbelin** [comment]
> `Definition F (id: forall [A] [x : A], A) := id.`

This is in `constrintern.ml` which does not know about typing and thus about what it means to be trailing. So I would renounce to do something at the current time.

**Comment by SimonBoulier** [comment]
I agree with Hugo. Hence I think that the PR is mergeable as is. Do you agree or do you think it is worth presenting at the next Coq Call? (tomorrow?)

File: doc/sphinx/language/gallina-extensions.rst
Comment (herbelin) [inline doc/sphinx/language/gallina-extensions.rst]: Thanks for contributing this section which was missing.

```suggestion
and the automatic declaration mode in on, the manual implicit arguments are added to the
```

File: doc/sphinx/language/gallina-extensions.rst
Comment (herbelin) [inline doc/sphinx/language/gallina-extensions.rst]:
```suggestion
automatically declared ones.
```
(??)

**Review by herbelin — APPROVED** [review]
Good to me, in line with the [comment](https://github.com/coq/coq/pull/11235#issuecomment-562713747) from @mattam82: "My point is that if non-maximal implicits are here to stay then we can see this as a welcome improvement of the situation" and the subsequent comments of the discussion.

**Comment by SimonBoulier** [comment]
Thanks for your suggestions @herbelin. I committed them and rebased on top of master.

**Comment by herbelin** [comment]
@SimonBoulier: this seems ready, right? Then, merging tomorrow.

**Comment by SimonBoulier** [comment]
Cool!

**Pull request merged**
